src/archive/src/main.rs

```diff
@@ -0,0 +1,394 @@
+            user_number,
+            timestamp,
+            log_index: idx as u64,
+        };
+
+        // the only way this expect can trigger is when the key size is wrong.
+        // On other failures (e.g. no more stable memory available) the underlying StableBTreeMap will panic directly.
+        index.insert(key, ()).expect("bug: key size mismatch");
+    })
+}
+
+#[query]
+fn get_logs(index: Option<u64>, limit: Option<u16>) -> Logs {
+    let num_entries = sanitize_limit(limit);
+
+    with_log(|log| {
+        let length = log.len();
+        let start_idx = match index {
+            None => length.saturating_sub(num_entries),
+            Some(idx) => idx as usize,
+        };
+
+        let next_idx = if start_idx + num_entries < length {
+            Some((start_idx + num_entries) as u64)
+        } else {
+            None
+        };
+
+        let mut entries = Vec::with_capacity(log.len().min(num_entries));
+        for idx in start_idx..start_idx + num_entries {
+            let entry = match log.get(idx) {
+                None => break,
+                Some(entry) => entry,
+            };
+            entries.push(Some(
+                candid::decode_one(&entry).expect("failed to decode log entry"),
+            ))
+        }
+        Logs { entries, next_idx }
+    })
+}
+
+#[query]
+fn get_user_logs(user_number: u64, cursor: Option<Cursor>, limit: Option<u16>) -> UserLogs {
+    let num_entries = sanitize_limit(limit);
+
+    with_user_index(|index| {
+        let iterator = match cursor {
+            None => index.range(user_number.to_le_bytes().to_vec(), None),
+            Some(Cursor::NextToken { next_token }) => index.range(next_token.into_vec(), None),
+            Some(Cursor::Timestamp { timestamp }) => index.range(
+                user_number.to_le_bytes().to_vec(),
+                Some(timestamp.to_le_bytes().to_vec()),
+            ),
+        };
+
+        with_log(|log| {
+            let mut entries: Vec<(UserIndexKey, Vec<u8>)> = iterator
+                .take(num_entries + 1) // take one too many to extract the cursor
+                .map(|(user_key, _)| {
+                    let entry = log
+                        .get(user_key.log_index as usize)
+                        .expect("bug: index to non-existing entry");
+                    (user_key, entry)
+                })
+                .collect();
+
+            let cursor = if entries.len() > num_entries {
+                entries.pop().map(|(key, _)| Cursor::NextToken {
+                    next_token: ByteBuf::from(key.to_bytes()),
+                })
+            } else {
+                None
+            };
+
+            let entries = entries
+                .iter()
+                .map(|(_, entry)| candid::decode_one(&entry).expect("failed to decode log entry"))
+                .collect();
+
+            UserLogs { entries, cursor }
+        })
+    })
+}
+
+fn sanitize_limit(limit: Option<u16>) -> usize {
```

Comment:
nit: I find it odd to have an `Option` as argument for something that's supposed to be "sanitizing" the input. I'd only expected a number or string to be sanitized. How about renaming (`limit_or_default` or I don't know what) or move everything unrelated to sanitization to the call site?